src/3d/terrain/qgsterraindownloader.cpp

```diff
@@ -0,0 +1,169 @@
+/***************************************************************************
+  qgsterraindownloader.cpp
+  --------------------------------------
+  Date                 : March 2019
+  Copyright            : (C) 2019 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsterraindownloader.h"
+
+#include "qgslogger.h"
+#include "qgsrasterlayer.h"
+
+#include "qgsgdalutils.h"
+
+
+QgsTerrainDownloader::QgsTerrainDownloader()
+{
+  QString uri = "type=xyz&url=http://s3.amazonaws.com/elevation-tiles-prod/terrarium/{z}/{x}/{y}.png&zmax=15&zmin=0";
+  onlineDtm.reset( new QgsRasterLayer( uri, "terrarium", "wms" ) );
+
+  // the whole world is projected to a square:
+  // X going from 180 W to 180 E
+  // Y going from ~85 N to ~85 S  (=atan(sinh(pi)) ... to get a square)
+  Q_NOWARN_DEPRECATED_PUSH
+  QgsCoordinateTransform ct( QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ), QgsCoordinateReferenceSystem( "EPSG:3857" ) );
+  Q_NOWARN_DEPRECATED_POP
+  QgsPointXY topLeftLonLat( -180, 180.0 / M_PI * std::atan( std::sinh( M_PI ) ) );
+  QgsPointXY bottomRightLonLat( 180, 180.0 / M_PI * std::atan( std::sinh( -M_PI ) ) );
+  QgsPointXY topLeft = ct.transform( topLeftLonLat );
+  QgsPointXY bottomRight = ct.transform( bottomRightLonLat );
+  mXSpan = ( bottomRight.x() - topLeft.x() );
+}
+
+QgsTerrainDownloader::~QgsTerrainDownloader() = default;
+
+
+void QgsTerrainDownloader::adjustExtentAndResolution( double mupp, const QgsRectangle &extentOrig, QgsRectangle &extent, int &res )
+{
+  double xMin = floor( extentOrig.xMinimum() / mupp ) * mupp;
+  double xMax = ceil( extentOrig.xMaximum() / mupp ) * mupp;
+
+  double yMin = floor( extentOrig.yMinimum() / mupp ) * mupp;
+  double yMax = ceil( extentOrig.yMaximum() / mupp ) * mupp;
+
+  extent = QgsRectangle( xMin, yMin, xMax, yMax );
+  res = round( ( xMax - xMin ) / mupp );
+}
+
+
+double QgsTerrainDownloader::findBestTileResolution( double requestedMupp )
+{
+  int zoom = 0;
+  for ( ; zoom <= 15; ++zoom )
+  {
+    double tileMupp = mXSpan / ( 256 * ( 1 << zoom ) );
+    if ( tileMupp <= requestedMupp )
+      break;
+  }
+
+  if ( zoom > 15 ) zoom = 15;
+  double finalMupp = mXSpan / ( 256 * ( 1 << zoom ) );
+  return finalMupp;
+}
+
+
+void QgsTerrainDownloader::tileImageToHeightMap( const QImage &img, QByteArray &heightMap )
```

Comment:
"fetch terrain tiles" algorithm in processing toolbox is something I had in mind and it should be straightforward to use QgsTerrainDownloader class for implementation of such algorithm.

Integration with WMS provider is something I have not thought of and I am not sure if I like that :-) My concern is that this conversion is very specific to the concrete tile format and there are other encodings as well. It would also require addition of a new "postprocessing" filter to the XYZ tiles support which really has specific use. Other problem is that we would need to move essentially the whole QgsTerrainDownloader implementation to WMS provider. Normal image tiles are simply stretched/scaled as needed, but with terrain tiles we must not do that - and that's the reason of the extra logic in the terrain downloader class. Resampling in RGB space creates artifacts in heightmaps, that's why in terrain downloader we fetch them in the native resolution, then convert and finally resample to destination resolution/CRS...

Reply:
Fair enough - if there's a tool to facilitate downloads of the tiles for use as a raster layer, that works for me.